src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,5426 @@
+	bool            *updated;
+};
+
+struct fd_dup2 {
+	int fd_src, fd_dest;
+	bool dest_closed;
+};
+
+/* working dir of current process */
+static char            cur_dir[DFS_MAX_PATH] = "";
+static bool            segv_handler_inited;
+/* Old segv handler */
+struct sigaction       old_segv;
+
+/* the flag to indicate whether initlization is finished or not */
+static bool            hook_enabled;
+static pthread_mutex_t lock_dfs;
+static pthread_mutex_t lock_fd;
+static pthread_mutex_t lock_dirfd;
+static pthread_mutex_t lock_mmap;
+static pthread_mutex_t lock_fd_dup2ed;
+
+/* store ! umask to apply on mode when creating file to honor system umask */
+static mode_t          mode_not_umask;
+
+static void
+finalize_dfs(void);
+static void
+update_cwd(void);
+
+struct dir_hdl {
+	d_list_t   entry;
+	dfs_obj_t *oh;
+	char       name[DFS_MAX_PATH];
+};
+
+static inline struct dir_hdl *
+hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dir_hdl, entry);
+}
+
+static bool
+key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return (strncmp(hdl->name, (const char *)key, DFS_MAX_PATH) == 0);
+}
+
+static void
+rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int             rc;
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	rc = dfs_release(hdl->oh);
+	if (rc == ENOMEM)
+		rc = dfs_release(hdl->oh);
+	if (rc && bLog)
+		D_WARN("dfs_release() failed: %s", strerror(rc));
+	D_FREE(hdl);
+}
+
+static bool
+rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	return true;
+}
+
+static uint32_t
+rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return d_hash_string_u32(hdl->name, strnlen(hdl->name, DFS_MAX_PATH));
+}
+
+static d_hash_table_ops_t hdl_hash_ops = {.hop_key_cmp    = key_cmp,
+					  .hop_rec_decref = rec_decref,
+					  .hop_rec_free   = rec_free,
+					  .hop_rec_hash   = rec_hash};
+
+/**
+ * Look up a path in hash table to get the dfs_obj for a dir.
+ * If not found in hash table, call to dfs_lookup() to open the object on DAOS.
+ * If the object is corresponding to a dir, insert the object into hash table.
+ * Since many DFS APIs need parent dir DFS object as parameter, we use a hash
+ * table to cache parent objects for efficiency.
+ */
+static int
+lookup_insert_dir(int idx_dfs, const char *name, mode_t *mode, dfs_obj_t **obj)
+{
+	struct dir_hdl *hdl;
+	dfs_obj_t      *oh;
+	d_list_t       *rlink;
+	size_t          len;
+	int             rc;
+
+	if (obj == NULL) {
+		D_ERROR("Invalid pointer!");
+		return EFAULT;
+	}
+	*obj = NULL;
+	len = strnlen(name, DFS_MAX_PATH);
+	if (len >= DFS_MAX_PATH) {
+		D_ERROR("path is too long!");
+		return ENAMETOOLONG;
+	}
+	D_RWLOCK_RDLOCK(&dfs_list[idx_dfs].lock_ht);
+	rlink = d_hash_rec_find(dfs_list[idx_dfs].dfs_dir_hash, name, len);
+	D_RWLOCK_UNLOCK(&dfs_list[idx_dfs].lock_ht);
```

Comment:
What are you trying to achieve with this lock?  It is technically needed as-is as the hash table isn't using the MUTEX flag so does not have locking but it does not ensure hash table consistency.  Please see my other comment on hash tables - the first step is working out what your requirements are from this HT.